### PR TITLE
Working Evolution Pathfinding

### PR DESCRIPTION
**Pull request by AarCon**
In order to call this, initialize it into a constant with:
```js
const evolutions = evolutionPathfinding()
```
Then you can get the specific pokemon's objects by calling it like this:
```js
console.log(evolutions['133']) //Eevee's PokemonID as an example
```
This will return the pokemon's data in an object:
```js
{
  path: [
    133, 470, 471,
    135, 134, 136,
    700, 196, 197
  ],
  method: [
    8,  8, 26, 27, 8, 8,
    8, 29,  2,  3, 8, 8,
    8
  ],
  ar: [
    [
        8,  85, 470,   0,   0,   8, 849, 471,   0,   0,  26,   0,
      470,   0,   0,  27,   0, 471,   0,   0,   8,  83, 135,   0,
        0,   8,  84, 134,   0,   0,   8,  82, 136,   0,   0,  29,
       17, 700,   0,   0,   2,   0, 196,   0,   0,   3,   0, 197,
        0,   0,   8, 107, 700,   0,   0,   8,  80, 196,   0,   0,
        8,  81, 197,   0,   0
    ],
    [],
    [],
    [],
    [],
    [],
    [],
    [],
    []
  ],
  targets: [
    470, 471, 135,
    134, 136, 700,
    196, 197
  ]
}
```
The `path` is the evolution path that is required for the pokemon. 
The `method` is ordered in the same way as `path` and details what method is used to evolve the original pokemon. 
The `ar` is the raw info from the EvolveTable.json and can be largely ignored. 
The `targets` is for every pokemon that the initial pokemon (Eevee) can evolve into the next stage. On the first form of a pokemon the `targets` will only show the next stage evolution. For example, Charmander will only target Charmeleon and Charmeleon will only target Charizard. Or in the Eevee example, it will evolve into all of the eeveelutions.

### CODE REVIEW

File: src/utils/dex/evolution.js [hunk 1]
```diff
@@ -235,6 +236,187 @@ function validIds(path, ar) {
...
 }
 
+function genForms() {
+  const formNamedata = formPokemonNames;
+  const formsList = formNamedata["labelDataArray"];
+  const forms = {};
+
+  for (let i = 0; i < formsList.length; i++) {
+    const allForms = formsList[i];
+    if (allForms["arrayIndex"] !== 0 && parseInt(allForms["labelName"].slice(-3)) > 0) {
```
Comment (Denmark14) [inline src/utils/dex/evolution.js:246]: There's a lot going on in this expression, can you extract it to a named variable?

File: src/utils/dex/evolution.js [hunk 2]
```diff
@@ -235,6 +236,187 @@ function validIds(path, ar) {
...
+  if (String(monsNo).length > 3) {
+    return `ZKN_FORM_$${monsNo}_${'0'.repeat(formZeros)}${formNo}`;
+  }
+  return `ZKN_FORM_${'0'.repeat(monZeros)}${monsNo}_${'0'.repeat(formZeros)}${formNo}`;
+}
+
+function removeDuplicates(pathDictionary) {
+  const newPath = [];
+  for (const pathElement of pathDictionary) {
+    if (!newPath.includes(pathElement)) {
```
Comment (Denmark14) [inline src/utils/dex/evolution.js:268]: If you want to ensure uniqueness, you can use a [Set](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set) instead

File: src/utils/dex/evolution.js [hunk 3]
```diff
@@ -235,6 +236,187 @@ function validIds(path, ar) {
...
+      newPath.push(pathElement);
+    }
+  }
+  return newPath;
+}
+
+function removeDuplicateForms(evolutionPaths) {
```
Comment (Denmark14) [inline src/utils/dex/evolution.js:275]: If you want to ensure uniqueness, you can use a [Set](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set) instead

File: src/utils/dex/evolution.js [hunk 4]
```diff
@@ -235,6 +236,187 @@ function validIds(path, ar) {
...
+}
+
+function updateEvolvePaths(evolutionPaths, currentMon, currentMonPath) {
+  evolutionPaths[currentMon].path.push(currentMon);
+  evolutionPaths[currentMon].path = [...new Set(evolutionPaths[currentMon].path)];
+
+  for (let i = 0; i < currentMonPath.length; i++) {
+    evolutionPaths[currentMonPath[i]].path.push(currentMon);
+    evolutionPaths[currentMonPath[i]].path = evolutionPaths[currentMonPath[i]].path.filter(
+      (x, j) => !new Set(evolutionPaths[currentMonPath[i]].path.slice(j + 1)).has(x)
```
Comment (Denmark14) [inline src/utils/dex/evolution.js:313]: Give x and j names please

**Review by Denmark14 — CHANGES_REQUESTED** [review]

**Comment by Denmark14** [comment]
Also apparently this PR broke a test?

File: src/utils/dex/evolution.js (hunk 2 above)
Comment (AarCon) [reply src/utils/dex/evolution.js:268]: I'm pretty sure this works. Let me know if if doesn't

**Review by Denmark14 — APPROVED** [review]

**Pull request merged**